### PR TITLE
add is-write-index flag to aliases

### PR DESCRIPTION
**Pull request by talevy**
This commit adds the is-write-index flag for aliases.
It allows requests to set the flag, and responses to display the flag.
It does not validate and/or affect any indexing/getting/updating behavior
of Elasticsearch -- this will be done in a follow-up PR.

This is a follow-up to: https://github.com/elastic/elasticsearch/pull/30703
It made more sense to split the above PR into two PRs. One for introducing the `is_write_index`
flag, and another to use it for validation and index resolution.

### CODE REVIEW

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.put_alias/all_path_options.yml [hunk 1]
```diff
@@ -27,9 +27,9 @@ setup:
       indices.get_alias:
         name: alias
 
-  - match: {test_index1.aliases.alias: {}}
+  - match: {test_index1.aliases.alias: { is_write_index: false }}
```
Comment (bleskes) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.put_alias/all_path_options.yml]: why  are these false? this is the only index in this alias so it should be true?

File: server/src/main/java/org/elasticsearch/action/admin/indices/alias/Alias.java [hunk 2]
```diff
@@ -60,6 +62,8 @@
     @Nullable
     private String searchRouting;
 
+    private boolean writeIndex;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/action/admin/indices/alias/Alias.java]: How do we specify "default"? don't we want a Boolean here?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 3]
```diff
@@ -516,7 +516,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                 }
                 for (Alias alias : request.aliases()) {
                     AliasMetaData aliasMetaData = AliasMetaData.builder(alias.name()).filter(alias.filter())
-                        .indexRouting(alias.indexRouting()).searchRouting(alias.searchRouting()).build();
+                        .indexRouting(alias.indexRouting()).searchRouting(alias.searchRouting()).writeIndex(alias.isWriteIndex()).build();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: we need to resolve the is write index here (for when it's not specified).

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasAction.java [hunk 4]
```diff
@@ -112,7 +121,8 @@ boolean removeIndex() {
         boolean apply(NewAliasValidator aliasValidator, MetaData.Builder metadata, IndexMetaData index) {
             aliasValidator.validate(alias, indexRouting, filter);
             AliasMetaData newAliasMd = AliasMetaData.newAliasMetaDataBuilder(alias).filter(filter).indexRouting(indexRouting)
-                    .searchRouting(searchRouting).build();
+                    .searchRouting(searchRouting).writeIndex(writeIndex).build();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/AliasAction.java]: what happens if writeIndex is null?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks @talevy , much smaller :) left some comments.

File: docs/reference/indices/aliases.asciidoc [hunk 5]
```diff
@@ -244,6 +244,65 @@ GET /alias2/_search?q=user:kimchy&routing=2,3
 // CONSOLE
 // TEST[continued]
 
+[float]
+[[aliases-write-index]]
+==== Write Index
+
+It is possible to associate the index pointed to by an alias as the write index.
+When specified, all index and update requests against an alias that point to multiple
```
Comment (bleskes) [inline docs/reference/indices/aliases.asciidoc]: I think that if you explicitly set a the write flag to false, we shouldn't write even if it's the only index the alias points to.

**Comment by talevy** [comment]
@bleskes thanks for the initial review. As I mentioned outside of Github, I believe we were misaligned in terms of how to break apart the original PR that both added the flag, and leveraged it to resolve index name expressions.

I've updated this PR to re-add validation and auto-upgrading of the `is_write_index`. This means that alias-index configurations with `is_write_index=null` in both alias-add-actions and index-creations will be set to `true` when there is no existing write index for that alias. In both cases that a promotion to `true` is not possible, `false` will be set.

Although AliasMetaData stores this value as a `Boolean`, any cluster-state changes that occur via index-creation or alias-updates will resolve this value to either `true` or `false`. AliasMetaData#writeIndex must stay a Boolean because of its use in templates. Templates need not define this field explicitly, and no smart unboxing should occur.

I've also added some unit tests that should account for these scenarios. (still iterating to catch all the CI tests that need updating).

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.put_alias/all_path_options.yml [hunk 6]
```diff
@@ -45,9 +45,9 @@ setup:
...
 
-  - match: {test_index1.aliases.alias: {}}
-  - match: {test_index2.aliases.alias: {}}
-  - match: {foo.aliases.alias: {}}
+  - match: {test_index1.aliases.alias: { is_write_index: true }}
+  - match: {test_index2.aliases.alias: { is_write_index: false }}
+  - match: {foo.aliases.alias: { is_write_index: false }}
 
 ---
 "put alias in * index":
```
Comment (talevy) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.put_alias/all_path_options.yml]: @bleskes, we discussed early about how it felt awkward to allow the defaulting of `is_write_index -> true` when 
not specified in alias updates apis (in addition to index creation). Towards the end of our discussion it seemed 
reasonable to rethink this and maybe enable this behavior. Although ordering was clear in the index-aliases api where actions are specified in a deterministic order, `indices.put_alias` allows resolution to multiple indices in such a way that ordering is undefined.

For these reasons, do you think it would be OK if we only default (when possible) `is_write_index -> true` during index creation?

**Comment by talevy** [comment]
@bleskes I've updated to reflect the changes we mentioned offline

- move all validation logic to run at the time that `MetaData` is built
- default `is_write_index (null)` to `true` if-and-only-if there are no other aliases, `false` otherwise

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasOrIndex.java [hunk 7]
```diff
@@ -134,8 +149,11 @@ public AliasMetaData getFirstAliasMetaData() {
             return referenceIndexMetaDatas.get(0).getAliases().get(aliasName);
         }
 
-        void addIndex(IndexMetaData indexMetaData) {
+        void addIndex(IndexMetaData indexMetaData, @Nullable Boolean isWriteIndex) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/AliasOrIndex.java]: Instead of doing this like this maybe only do it at the end? add a method called "computeAndValidateWriteIndex" and let that one do two things:
- scan the all index metadata and extracts the writeIndex, if any
- validate we have no two indices which are marked with the index_write flags.

you can use org.apache.lucene.util.SetOnce to store the writeIndex reference. Then you can assert that once `computeAndValidateWriteIndex` is called you can't call addIndex anymore.

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasValidator.java [hunk 8]
```diff
@@ -145,6 +148,29 @@ public void validateAliasFilter(String alias, byte[] filter, QueryShardContext q
         }
     }
 
+    /**
+     * Validates that there is only one write index per alias
+     *
+     * @param aliasAndIndexLookup The current state of aliases and indices in the cluster state
+     * @throws IllegalStateException if any alias has more than one write index
+     */
+    public static void validateAliasWriteOnly(Map<String, AliasOrIndex> aliasAndIndexLookup) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/AliasValidator.java]: see my other comment.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 9]
```diff
@@ -701,7 +702,8 @@ private static IndexNotFoundException indexNotFoundException(String expression)
                 if (context.getOptions().ignoreAliases()) {
                     return metaData.getAliasAndIndexLookup().entrySet().stream()
                             .filter(e -> e.getValue().isAlias() == false)
-                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                                (v1, v2) -> {throw new IllegalStateException("no duplicate indices should exist");}, TreeMap::new));
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: out of curiosity - why did you add this?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 10]
```diff
@@ -557,6 +557,25 @@ public IndexMetaData getIndexSafe(Index index) {
         throw new IndexNotFoundException(index);
     }
 
+    /**
+     * Retrieves the write-index for a given <code>alias</code>
+     *
+     * @param alias the alias to search for a write-index with
+     * @return the write-index {@link Index} for <code>alias</code>, null if no write-index is configured
+     */
+    public Index getWriteIndex(String alias) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: this is usage? I think we should remove it and use it in the follow up PRs we're planning?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 11]
```diff
@@ -1040,6 +1059,34 @@ public MetaData build() {
             }
 
             // build all indices map
+
+            SortedMap<String, AliasOrIndex> aliasAndIndexLookup = buildAliasAndIndexLookup();
+            AliasValidator.validateAliasWriteOnly(aliasAndIndexLookup);
+
+            // set any `is_write_index == null` to their appropriate boolean values
+            for (AliasOrIndex aliasOrIndex : aliasAndIndexLookup.values()) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: see other comment - this should all be done in buildAliasAndIndexLookup

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 12]
```diff
@@ -1040,6 +1059,34 @@ public MetaData build() {
...
+                    AliasOrIndex.Alias alias = (AliasOrIndex.Alias) aliasOrIndex;
+                    updateIndicesWithDefaultWriteIndex(alias);
+                }
+            }
+
+            aliasAndIndexLookup = Collections.unmodifiableSortedMap(buildAliasAndIndexLookup());
+
+
+            // build all concrete indices arrays:
+            // TODO: I think we can remove these arrays. it isn't worth the effort, for operations on all indices.
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I believe this is about memory allocation - I think we can discuss this but in a different issue.

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx @talevy . I left some suggestions.

File: docs/reference/indices/aliases.asciidoc [hunk 13]
```diff
@@ -244,6 +244,94 @@ GET /alias2/_search?q=user:kimchy&routing=2,3
...
+                 "is_write_index" : false
+            }
+        }
+    ]
+}
+--------------------------------------------------
+// CONSOLE
+// TEST[s/^/PUT test\nPUT test2\n/]
+
+It is important to note that the order of these actions is important. The original write index
```
[the remark this answers is not in the document]
Comment (talevy) [reply docs/reference/indices/aliases.asciidoc]: oh, woops. I thought I went through this. will update since it is left-over and currently wrong

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java [hunk 14]
```diff
@@ -308,6 +337,8 @@ public static void toXContent(AliasMetaData aliasMetaData, XContentBuilder build
                 builder.field("search_routing", aliasMetaData.searchRouting());
             }
 
+            builder.field("is_write_index", Boolean.TRUE.equals(aliasMetaData.writeIndex()));
```
[the remark this answers is not in the document]
Comment (talevy) [reply server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java]: The aspect of Clint's original proposal that I liked was that we would always show this field and make it clear to the user which value we defaulted to. Having it be `null` feels confusing.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 9 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: oh woops. this was not meant to stay. I added this to debug where the ordering of indices was being lost. will revert

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java]: > make it clear to the user which value we defaulted to

That's my point - we don't know in this point. I agree that outputting `null` will be confusing hence my suggestion which will completely omit the field if not set.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 10 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I will double check, but I believe it will be used by the index-name resolver when used for indexing. It is not used yet since we haven't implemented that yet

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 12 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I didn't write this

**Comment by talevy** [comment]
hey @bleskes, I've updated the build logic to behave more like a one-pass build of the aliasAndIndexLookup + manipulating the IndexMetaData of the cluster, itself.

one thing I am debugging is how this behavior is affecting snapshot restoration. Since one can restore and rename index names, the aliasmetadata now has multiple copies of `is_write_index: true`. So [restoration is failing](https://github.com/talevy/elasticsearch/blob/bdb0fb25555381299d80fed7efcbcab97a9d172f/server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java#L1540)

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java [hunk 15]
```diff
@@ -194,6 +210,11 @@ public AliasMetaData(StreamInput in) throws IOException {
...
             searchRoutingValues = emptySet();
         }
+        if (in.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
+            writeIndex = in.readOptionalBoolean();
+        } else {
+            writeIndex = null;
+        }
     }
 
     public static Diff<AliasMetaData> readDiffFrom(StreamInput in) throws IOException {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java]: I have a general question about cluster-state diffs, if you don't mind sharing. We seem to check stream-input version in the general Writeable.Reader, but I do not see any here. I'm not entirely sure why that is the case, shouldn't we also version-guard reading diffs across the wire?

Comment (talevy) [reply server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java]: oh, nevermind. the diff is just a complete overwrite. not part-by-part, so the wire-version is covered

File: server/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequestTests.java [hunk 16]
```diff
@@ -93,6 +93,7 @@ public void testToXContent() throws IOException {
         Alias alias = new Alias("test_alias");
         alias.routing("1");
         alias.filter("{\"term\":{\"year\":2016}}");
+        alias.writeIndex(true);
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequestTests.java]: randomize?

File: server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java [hunk 17]
```diff
@@ -95,6 +111,7 @@ private static AliasMetaData createTestItem() {
         if (randomBoolean()) {
             builder.filter("{\"term\":{\"year\":2016}}");
         }
+        builder.writeIndex(randomBoolean());
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java]: sometime make it null?

File: server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java [hunk 18]
```diff
@@ -54,6 +55,21 @@ public void testSerialization() throws IOException {
...
 
+    @Override
+    protected void assertEqualInstances(AliasMetaData expectedInstance, AliasMetaData newInstance) {
+        assertNotSame(newInstance, expectedInstance);
+        if (expectedInstance.writeIndex() == null) {
+            expectedInstance = AliasMetaData.builder(expectedInstance.alias())
+                .filter(expectedInstance.filter())
+                .indexRouting(expectedInstance.indexRouting())
+                .searchRouting(expectedInstance.searchRouting())
+                .writeIndex(Boolean.FALSE)
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java]: why is that force false?

File: server/src/test/java/org/elasticsearch/cluster/metadata/AliasValidatorTests.java [hunk 19]
```diff
@@ -44,4 +45,6 @@ public void testValidatesAliasNames() {
         // Doesn't throw an exception because we allow upper case alias names
         validator.validateAliasStandalone("CAT", null);
     }
+
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/metadata/AliasValidatorTests.java]: left overs

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java [hunk 20]
```diff
@@ -291,6 +296,32 @@ public void testValidateWaitForActiveShardsFailure() throws Exception {
         assertThat(e.getMessage(), containsString("invalid wait_for_active_shards"));
     }
 
+    public void testWriteIndex() throws Exception {
+        Boolean writeIndex = randomBoolean();
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java]: random null?

File: server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java [hunk 21]
```diff
@@ -111,7 +111,7 @@ public void testSimpleJsonFromAndTo() throws IOException {
                         .putMapping("mapping1", MAPPING_SOURCE1)
                         .putMapping("mapping2", MAPPING_SOURCE2)
                         .putAlias(newAliasMetaDataBuilder("alias1").filter(ALIAS_FILTER1))
-                        .putAlias(newAliasMetaDataBuilder("alias2"))
+                        .putAlias(newAliasMetaDataBuilder("alias2").writeIndex(true))
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java]: some randomization comments

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexAliasesServiceTests.java [hunk 22]
```diff
@@ -125,6 +128,63 @@ public void testRemoveIndexTwice() {
...
+            equalTo(after.metaData().index("test")));
+
+        after = service.innerExecute(before, Arrays.asList(
+            new AliasAction.Add("test", "alias", null, null, null, true)));
+        assertTrue(after.metaData().index("test").getAliases().get("alias").writeIndex());
+        assertThat(((AliasOrIndex.Alias) after.metaData().getAliasAndIndexLookup().get("alias")).getWriteIndex(),
+            equalTo(after.metaData().index("test")));
+    }
+
+    public void testAddWriteOnlyWithExistingWriteIndex() {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexAliasesServiceTests.java]: I think we need more tests here - adding a write index to a group of non write indexes. removing an index from a default group of two indices (and seeing there is a write index)

**Review by bleskes — CHANGES_REQUESTED** [review]
This looks great. I left some nits on the testing.

File: server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java (hunk 17 above)
Comment (talevy) [reply server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java]: I think it is ok to leave it as static since that is the nature of this test.

I believe testToAndFromXContent captures the random scenario, and so I will update it so
that randomAlias includes the write index randomization

File: server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java (hunk 18 above)
Comment (talevy) [reply server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java]: just leaving the static nature of this test, will update

File: server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java (hunk 21 above)
Comment (talevy) [reply server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java]: as before, I chose to leave the static nature of this test. I can update here

**Comment by talevy** [comment]
I've updated to add more tests and removed unused imports, ready for another review @bleskes. thanks!

**Review by bleskes — APPROVED** [review]
LGTM. looks great  @talevy

**Comment by talevy** [comment]
thanks! I will probably finalize the back-porting dance early next week in the hopes of not breaking a CI run on a Friday

**Comment by talevy** [comment]
thanks @albertzaharovits! sorry for the noise
